packages/effect/src/internal/stream.ts

```diff
@@ -5709,42 +5710,49 @@ export const shareRefCount = dual<
     readonly connector?: Effect.Effect<PubSub.PubSub<Take.Take<A, E>>>
   }) => <R>(
     self: Stream.Stream<A, E, R>
-  ) => Stream.Stream<A, E, R>,
+  ) => Effect.Effect<Stream.Stream<A, E, R>, never, R | Scope.Scope>,
   <A, E, R>(
     self: Stream.Stream<A, E, R>,
     config: {
       readonly connector?: Effect.Effect<PubSub.PubSub<Take.Take<A, E>>>
     }
-  ) => Stream.Stream<A, E, R>
+  ) => Effect.Effect<Stream.Stream<A, E, R>, never, R | Scope.Scope>
 >(
   2,
   <A, E, R>(
     self: Stream.Stream<A, E, R>,
     config: {
       readonly connector?: Effect.Effect<PubSub.PubSub<Take.Take<A, E>>>
     }
-  ): Stream.Stream<A, E, R> => {
-    let refCount = 0
-    let connector: PubSub.PubSub<Take.Take<A, E>> | null = null
-    let fiber: Fiber.RuntimeFiber<void> | null = null
+  ): Effect.Effect<Stream.Stream<A, E, R>, never, R | Scope.Scope> => {
     return Effect.gen(function*() {
-      refCount++
-      connector ??= yield* (config.connector ?? PubSub.bounded<Take.Take<A, E>>(16))
-      fiber ??= yield* Effect.forkDaemon(runIntoPubSub(self, connector))
-      return flattenTake(fromPubSub(connector))
-    }).pipe(
-      unwrap,
-      ensuring(
-        Effect.suspend(() => {
-          refCount--
-          const cleanup = fiber
-          fiber = null
-          return refCount === 0 && cleanup
-            ? Fiber.interrupt(cleanup)
-            : Effect.void
-        })
+      let refCount = 0
+      let connector: PubSub.PubSub<Take.Take<A, E>> | null = null
+      let fiber: Fiber.RuntimeFiber<void> | null = null
+      const scope = yield* Effect.scope
+      const context = yield* Effect.context<R>()
+      return Effect.gen(function*() {
+        refCount++
+        connector ??= yield* (config.connector ?? PubSub.bounded<Take.Take<A, E>>(16))
+        fiber ??= yield* runIntoPubSub(self, connector).pipe(
+          Effect.locally(FiberRef.currentContext, context),
+          Effect.forkIn(scope)
+        )
+        return flattenTake(fromPubSub(connector))
+      }).pipe(
+        unwrap,
+        ensuring(
+          Effect.suspend(() => {
+            refCount--
+            const cleanup = fiber
+            fiber = null
```

Comment:
You are clearing the fiber even if the ref count is not 0.

Reply:
it doesn't matter what is the ref count, theres a new pubsub and new forked fiber for every evaluation of listening

Reply:
I use [Nullish coalescing assignment](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Nullish_coalescing_assignment), so i don't think it's the case here

Reply:
You're right, only a new forked fiber on every eval, not new `PubSub`, I glazed over the `??=`

You should still instantiate the `PubSub` earlier imo. Theres no reason that a pubsub should be instantiated as a hidden side-effect of listening to a stream